doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,189 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to report diagnostic data
+via named channels. It can be accessed using:
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+```
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Subscribe to a channel by name
+diagnostics_channel.subscribe('my-channel', (name, data) => {
+  // Received data
+});
+
+// Subscribe to any channels matching a function result
+diagnostics_channel.subscribe({
+  shouldListen(name) {
+    return /^\s+-channel/.test(name);
+  },
+  onMessage(name, data) {
+    // Received data
+  }
+});
+
+// Subscribe to every channel
+diagnostics_channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Get a reusable channel object
+const channel = diagnostics_channel.publisher('my-channel');
+
+// Subscribe directly to a channel object
+channel.subscribe((name, data) => {
+  // Received data
+});
+
+// Check if the channel has an active subscriber
+if (channel.shouldPublish()) {
+  // Publish data to the channel
+  channel.publish({
+    some: 'data'
+  });
+}
+
+// An alternative way to check for active subsribers before publishing
+channel.ifActive?.publish('my-channel', {
+  some: 'data'
+});
+```
+
+#### `diagnostics_channel.subscribe([name, ] onMessage)`
+
+* `name` {string} The channel name
+* `onMessage` {Function} The handler to receive channel messages
```

Comment:
Please add some more info regarding the arguments of `onMessage`.
We should also define what happens in case `onMessage` throws.

Reply:
Closely related... We should definitely what happens when an async function is provided here and it rejects.

Reply:
I would think it should just become an unhandled rejection. I added a change so a regular function throwing an error will trigger uncaught exception so it should handle errors in a reasonable way. I'll make sure to add a note to the doc about that behaviour too.